File: scene/src/main/java/com/gx/kotlin/scene/roam/NativeRoamView.java
package com.gx.kotlin.scene.roam;

import android.content.Context;
import android.graphics.PointF;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.gx.kotlin.scene.camera.CameraPosition;
import com.gx.kotlin.scene.geometry.WayPoint;
import com.gx.kotlin.scene.geometry.WayPointBounds;
import com.gx.kotlin.scene.roam.renderer.RoamRenderer;

public final class NativeRoamView implements NativeRoam {

    private final RoamRenderer mRoamRenderer;

    // Device density
    private final float pixelRatio;

    private long nativePtr = 0;

    public NativeRoamView(final Context context, final RoamRenderer roamRenderer) {
        this.pixelRatio = context.getResources().getDisplayMetrics().density;
        this.mRoamRenderer = roamRenderer;
        nativeInitialize(this, mRoamRenderer, pixelRatio);
    }

    @Override
    public void moveBy(double deltaX, double deltaY, long duration) {
        nativeMoveBy(deltaX, deltaY, duration);
    }

    @Override
    public double getBearing() {
        return nativeGetBearing();
    }

    @Override
    public void setBearing(double degrees, double fx, double fy, long duration) {
        nativeSetBearingXY(degrees, fx / pixelRatio, fy / pixelRatio, duration);
    }

    @Override
    public double getPitch() {
        return nativeGetPitch();
    }

    @Override
    public void setLatLngBounds(@Nullable WayPointBounds wayPointBounds) {
        nativeSetWayPointBounds(wayPointBounds);
    }

    @Override
    public void setPitch(double pitch, long duration) {
        nativeSetPitch(pitch, duration);
    }

    @Override
    public double getZoom() {
        return nativeGetZoom();
    }

    @Override
    public void setZoom(double zoom, @NonNull PointF focalPoint, long duration) {
        nativeSetZoom(zoom, focalPoint.x / pixelRatio, focalPoint.y / pixelRatio, duration);
    }

    @Override
    public void setMinZoom(double zoom) {
        nativeSetMinZoom(zoom);
    }

    @Override
    public double getMinZoom() {
        return nativeGetMinZoom();
    }

    @Override
    public void setMaxZoom(double zoom) {
        nativeSetMaxZoom(zoom);
    }

    @Override
    public double getMaxZoom() {
        return nativeGetMaxZoom();
    }


    @Override
    public void resizeView(int width, int height) {
        width =  (int) Math.ceil(width / pixelRatio);
        height = (int) Math.ceil(height / pixelRatio);

        nativeResizeView(width, height);
    }

    @Override
    public void setWayPoint(@NonNull WayPoint wayPoint, long duration) {
        nativeSetWayPoint(wayPoint.getX(), wayPoint.getY(), duration);
    }

    @Override
    public CameraPosition getCameraPosition() {
        return nativeGetCameraPosition();
    }

    @Override
    public CameraPosition getCameraForWayPointBounds(WayPointBounds bounds, int[] padding, double bearing, double tilt) {
        return nativeGetCameraForWayPointBounds(
                bounds,
                padding[1] / pixelRatio,
                padding[0] / pixelRatio,
                padding[3] / pixelRatio,
                padding[2] / pixelRatio,
                bearing,
                tilt
        );
    }

    @Override
    public void jumpTo(@NonNull WayPoint center, double zoom, double pitch, double angle) {
        nativeJumpTo(angle, center.getX(), center.getY(), pitch, zoom);
    }

    @Override
    public void easeTo(@NonNull WayPoint center, double zoom, double angle, double pitch, long duration,
                       boolean easingInterpolator) {
        nativeEaseTo(angle, center.getX(), center.getY(), duration, pitch, zoom,
                easingInterpolator);
    }

    @NonNull
    @Override
    public PointF pixelForWayPoint(@NonNull WayPoint wayPoint) {
        PointF pointF = nativePixelForWayPoint(wayPoint.getX(), wayPoint.getY());
        pointF.set(pointF.x * pixelRatio, pointF.y * pixelRatio);
        return pointF;
    }

    @Override
    public WayPoint wayPointForPixel(@NonNull PointF pixel) {
        return nativeWayPointForPixel(pixel.x / pixelRatio, pixel.y / pixelRatio);
    }


    private native void nativeInitialize(NativeRoamView nativeRoamView, RoamRenderer mapRenderer, float pixelRatio);

    private native void nativeDestroy();

    private native void nativeResizeView(int width, int height);

    private native void nativeMoveBy(double deltaX, double deltaY, long duration);

    private native void nativeSetBearingXY(double degrees, double fx, double fy, long duration);

    private native double nativeGetBearing();

    private native double nativeGetPitch();

    private native void nativeSetPitch(double pitch, long duration);

    private native void nativeSetZoom(double zoom, double cx, double cy, long duration);

    private native double nativeGetZoom();

    private native void nativeSetMinZoom(double zoom);

    private native double nativeGetMinZoom();

    private native void nativeSetMaxZoom(double zoom);

    private native double nativeGetMaxZoom();

    private native void nativeSetWayPoint(double wayPointX, double wayPointY, long duration);

    private native CameraPosition nativeGetCameraPosition();

    private native void nativeJumpTo(double angle, double x, double y, double pitch, double zoom);

    private native void nativeEaseTo(double angle, double x, double y,
                                     long duration, double pitch, double zoom,
                                     boolean easingInterpolator);

    private native void nativeSetWayPointBounds(WayPointBounds latLngBounds);

    private native CameraPosition nativeGetCameraForWayPointBounds(
            WayPointBounds wayPointBounds, double top, double left, double bottom, double right, double bearing, double tilt);

    private native PointF nativePixelForWayPoint(double x, double y);

    private native WayPoint nativeWayPointForPixel(float x, float y);


}

File: scene/src/main/cpp/native_roam_view.cpp
//
// Created by 龚喜 on 2024/8/19.
//

#include "native_roam_view.h"
#include "RenderLogger.h"
#include "roam_renderer.h"
#include "android_renderer_frontend.h"
#include "roam/roam_options.h"
#include "GLCamera.h"
#include <cmath>


NativeRoamView::NativeRoamView(jni::JNIEnv& _env,
                               const jni::Object<NativeRoamView>& obj,
                               const jni::Object<RoamRenderer>& jMapRender,
                               jni::jfloat _pixelRatio)
                               :mRoamRenderer(RoamRenderer::getNativePeer(_env, jMapRender))
                               ,mPixelRatio(_pixelRatio)
{
    mRendererFrontend = std::make_unique<AndroidRendererFrontend>(mRoamRenderer);

    roamgl::RoamOptions options;
    options.withSize(roamgl::Size{ static_cast<uint32_t>(mWidth), static_cast<uint32_t>(mHeight) })
    .withPixelRatio(mPixelRatio);

    mRoam = std::make_unique<roamgl::Roam>(*mRendererFrontend, options);

//    GLCamera camera;
//    mRoamRenderer.update(std::make_shared<GLCamera>(std::move(camera)));
}

NativeRoamView::~NativeRoamView()
{

}

void NativeRoamView::registerNative(JNIEnv& env) {
    // Lookup the class
    static auto& javaClass = jni::Class<NativeRoamView>::Singleton(env);


#define METHOD(MethodPtr, name) jni::MakeNativePeerMethod<decltype(MethodPtr), (MethodPtr)>(name)

    // Register the peer
    jni::RegisterNativePeer<NativeRoamView>(env, javaClass, "nativePtr",
                                            jni::MakePeer<NativeRoamView, const jni::Object<NativeRoamView>&, const jni::Object<RoamRenderer>&, jni::jfloat>,
                                            "nativeInitialize", "nativeDestroy",
                                            METHOD(&NativeRoamView::resizeView, "nativeResizeView"),
                                            METHOD(&NativeRoamView::moveBy, "nativeMoveBy"),
                                            METHOD(&NativeRoamView::easeTo, "nativeEaseTo"),
                                            METHOD(&NativeRoamView::setBearingXY, "nativeSetBearingXY"),
                                            METHOD(&NativeRoamView::getBearing, "nativeGetBearing"),
                                            METHOD(&NativeRoamView::setPitch, "nativeSetPitch"),
                                            METHOD(&NativeRoamView::getPitch, "nativeGetPitch"),
                                            METHOD(&NativeRoamView::getZoom, "nativeGetZoom"),
                                            METHOD(&NativeRoamView::setZoom, "nativeSetZoom"),
                                            METHOD(&NativeRoamView::setMinZoom, "nativeSetMinZoom"),
                                            METHOD(&NativeRoamView::getMinZoom, "nativeGetMinZoom"),
                                            METHOD(&NativeRoamView::setMaxZoom, "nativeSetMaxZoom"),
                                            METHOD(&NativeRoamView::getMaxZoom, "nativeGetMaxZoom"),
                                            METHOD(&NativeRoamView::getCameraPosition, "nativeGetCameraPosition"),
                                            METHOD(&NativeRoamView::jumpTo, "nativeJumpTo"),
                                            METHOD(&NativeRoamView::setWayPoint, "nativeSetWayPoint"),
                                            METHOD(&NativeRoamView::setWayPointBounds, "nativeSetWayPointBounds"),
                                            METHOD(&NativeRoamView::getCameraForWayPointBounds, "nativeGetCameraForWayPointBounds"),
                                            METHOD(&NativeRoamView::pixelForWayPoint, "nativePixelForWayPoint"),
                                            METHOD(&NativeRoamView::wayPointForPixel, "nativeWayPointForPixel")
                                            );
}


void NativeRoamView::resizeView(jni::JNIEnv&, int w, int h) {
    mWidth = std::max(64, w);
    mHeight = std::max(64, h);
    mRoam->setSize({ static_cast<uint32_t>(mWidth), static_cast<uint32_t>(mHeight) });
}


void NativeRoamView::moveBy(jni::JNIEnv&, jni::jdouble dx, jni::jdouble dy, jni::jlong duration) {
    LOGI("NativeRoamView moveBy %.2f %.2f", dx, dy);
    mRoam->moveBy({dx, dy});
}

void NativeRoamView::easeTo(jni::JNIEnv&, jni::jdouble bearing, jni::jdouble x, jni::jdouble y, jni::jlong duration, jni::jdouble pitch, jni::jdouble zoom, jni::jboolean easing) {
    roamgl::CameraOptions cameraOptions;
    if (bearing != -1) {
        cameraOptions.bearing = bearing;
    }
    cameraOptions.center = roamgl::WayPoint(x, y);
    cameraOptions.padding = insets;
    if (pitch != -1) {
        cameraOptions.pitch = pitch;
    }
    if (zoom != -1) {
        cameraOptions.zoom = zoom;
    }

    roamgl::AnimationOptions animationOptions;
    animationOptions.duration.emplace(roamgl::Milliseconds(duration));
    if (!easing) {
        // add a linear interpolator instead of easing
        animationOptions.easing.emplace(roamgl::util::UnitBezier { 0, 0, 1, 1 });
    }

    mRoam->easeTo(cameraOptions, animationOptions);
}

void NativeRoamView::setBearingXY(jni::JNIEnv&, jni::jdouble degrees, jni::jdouble cx, jni::jdouble cy, jni::jlong duration) {
    LOGI("NativeRoamView setBearingXY %.2f", degrees);
    roamgl::ScreenCoordinate anchor(cx, cy);
    mRoam->easeTo(roamgl::CameraOptions().withBearing(degrees).withAnchor(anchor), roamgl::AnimationOptions{roamgl::Milliseconds(duration)});
}

jni::jdouble NativeRoamView::getBearing(jni::JNIEnv&) {
    return *mRoam->getCameraOptions().bearing;
}

jni::jdouble NativeRoamView::getPitch(jni::JNIEnv&) {
    return *mRoam->getCameraOptions().pitch;
}

void NativeRoamView::setPitch(jni::JNIEnv&, jni::jdouble pitch, jni::jlong duration) {
    LOGI("NativeRoamView setPitch %.2f ", pitch);
    mRoam->easeTo(roamgl::CameraOptions().withPitch(pitch), roamgl::AnimationOptions{roamgl::Milliseconds(duration)});
}

void NativeRoamView::setZoom(jni::JNIEnv&, jni::jdouble zoom, jni::jdouble x, jni::jdouble y, jni::jlong duration) {
    LOGI("NativeRoamView setZoom %.2f [%.2f, %.2f]", zoom, x, y);
    mRoam->easeTo(roamgl::CameraOptions().withZoom(zoom).withAnchor(roamgl::ScreenCoordinate{ x, y }), roamgl::AnimationOptions{roamgl::Milliseconds(duration)});
}

jni::jdouble NativeRoamView::getZoom(jni::JNIEnv&) {
    return *(mRoam->getCameraOptions().zoom);
}

void NativeRoamView::setMinZoom(jni::JNIEnv&, jni::jdouble zoom)
{
    mRoam->setBounds(roamgl::BoundOptions().withMinZoom(zoom));
}

jni::jdouble NativeRoamView::getMinZoom(jni::JNIEnv&)
{
    return *mRoam->getBounds().minZoom;
}

void NativeRoamView::setMaxZoom(jni::JNIEnv&, jni::jdouble zoom)
{
    mRoam->setBounds(roamgl::BoundOptions().withMaxZoom(zoom));
}

jni::jdouble NativeRoamView::getMaxZoom(jni::JNIEnv&)
{
    return *mRoam->getBounds().maxZoom;
}

jni::Local<jni::Object<android::CameraPosition>> NativeRoamView::getCameraPosition(jni::JNIEnv& env)
{
    return android::CameraPosition::New(env, mRoam->getCameraOptions(insets));
}

void NativeRoamView::jumpTo(jni::JNIEnv&, jni::jdouble bearing, jni::jdouble wayPointX, jni::jdouble wayPointY, jni::jdouble pitch, jni::jdouble zoom)
{
    roamgl::CameraOptions options;
    if (bearing != -1) {
        options.bearing = bearing;
    }
    options.center = roamgl::WayPoint(wayPointX, wayPointY);
    if (pitch != -1) {
        options.pitch = pitch;
    }
    if (zoom != -1) {
        options.zoom = zoom;
    }

//    map->jumpTo(options);

}

void NativeRoamView::setWayPoint(jni::JNIEnv&, jni::jdouble wayPointX, jni::jdouble wayPointY, jni::jlong duration) {
    LOGI("setWayPoint setWayPoint %.2f %.2f ", wayPointX, wayPointY);
    mRoam->easeTo(roamgl::CameraOptions().withCenter(roamgl::WayPoint(wayPointX, wayPointY)).withPadding(insets), roamgl::AnimationOptions{roamgl::Milliseconds(duration)});
}

void NativeRoamView::setWayPointBounds(jni::JNIEnv& env, const jni::Object<android::WayPointBounds>& jBounds) {
    roamgl::BoundOptions bounds;
    if (jBounds) {
        bounds.withWayPointBounds(android::WayPointBounds::getWayPointBounds(env, jBounds));
    } else {
        bounds.withWayPointBounds(roamgl::WayPointBounds::unbounded());
    }
    mRoam->setBounds(bounds);
}

jni::Local<jni::Object<android::CameraPosition>> NativeRoamView::getCameraForWayPointBounds(jni::JNIEnv& env, const jni::Object<android::WayPointBounds>& jBounds, double top, double left, double bottom, double right, double bearing, double tilt) {
    roamgl::EdgeInsets padding = {top, left, bottom, right};
    return android::CameraPosition::New(env, mRoam->cameraForWayPointBounds(android::WayPointBounds::getWayPointBounds(env, jBounds), padding, bearing, tilt));
}


jni::Local<jni::Object<android::PointF>> NativeRoamView::pixelForWayPoint(JNIEnv& env, jdouble x, jdouble y)
{
    roamgl::ScreenCoordinate pixel = mRoam->pixelForWayPoint(roamgl::WayPoint(x, y));
    return android::PointF::New(env, static_cast<float>(pixel.x), static_cast<float>(pixel.y));
}


jni::Local<jni::Object<android::WayPoint>> NativeRoamView::wayPointForPixel(JNIEnv& env, jfloat x, jfloat y)
{
    return android::WayPoint::New(env, mRoam->wayPointForPixel(roamgl::ScreenCoordinate(x, y)));
}
